### PR TITLE
FAT-131 - Custom image tool

### PR DESCRIPTION
**Pull request by ofreyssinet-ledger**
### 📝 Description

Tool to import an image (either from the phone's gallery or from an URL) and display a cropping UI, then resize it to a target size, pick & apply a contrast value, preview the result and output the raw data.

New feature flag: `"customImage"`

**Two entry points:**
- in the NFT viewer, by pressing the "..." button, there is a "Custom Image" button (enabled by featureflag
- in debug settings

This is the first part of work for a bigger feature. Ask me on slack for more details about the what/why.

### ❓ Context

- **Impacted projects**: `ledger-live-mobile` `ledger-live-common` 
- **Linked resource(s)**: [FAT-135] [FAT-131] 

### ✅ Checklist

- [ ] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 📸 Demo

I can send you a video on Slack

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

[FAT-131]: https://ledgerhq.atlassian.net/browse/FAT-131?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[FAT-135]: https://ledgerhq.atlassian.net/browse/FAT-135?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/ledger-live-common/src/featureFlags/types.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ export type FeatureId =
   | "ratings"
   | "counterValue"
   | "buyDeviceFromLive"
-  | string;
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/featureFlags/types.ts:9]: leaving string was not strict enough in my opinion, because of that for instance we didn't detect that a default config for llmUsbFirmwareUpdate was missing

File: libs/ledger-live-common/src/featureFlags/defaultFeatures.ts [hunk 2]
```diff
@@ -64,4 +64,10 @@ export const defaultFeatures: DefaultFeatures = {
   counterValue: {
     enabled: false,
   },
+  llmUsbFirmwareUpdate: {
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/featureFlags/defaultFeatures.ts:67]: this was missing because the typing was not strict enough.
see https://github.com/LedgerHQ/ledger-live/pull/743/files#r930990175

File: apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import React, { useMemo } from "react";
+import { createStackNavigator } from "@react-navigation/stack";
+import { useTheme } from "styled-components/native";
+import { ScreenName } from "../../const";
+import { getStackNavigatorConfig } from "../../navigation/navigatorConfig";
+import Step1Cropping from "../../screens/CustomImage/Step1Crop";
+import Step2Preview from "../../screens/CustomImage/Step2Preview";
+import DebugScreen from "../../screens/CustomImage/DebugScreen";
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx:8]: `DebugScreen` is not used it looks like ?

File: apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      <Stack.Screen
+        name={ScreenName.CustomImageErrorScreen}
+        component={ErrorScreen}
+        options={{ title: "", headerLeft: Empty }}
+      />
+    </Stack.Navigator>
+  );
+}
+
+const Stack = createStackNavigator();
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx:47]: Maybe with new navigators we should enforce typing ? (only CustomImageNavigator, not the BaseNavigator)

You can find an example on Ledger Live (not yet merged) [here](https://github.com/LedgerHQ/ledger-live/blob/aa76c034a6400add49b20d79aba9089681c235f7/apps/ledger-live-mobile/src/components/RootNavigator/SyncOnboardingNavigator.tsx#L21) - and the [doc](https://reactnavigation.org/docs/typescript/)

File: apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx (hunk 3 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx:8]: indeed, it doesn't exist anymore actually

File: apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx (hunk 4 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx:47]: Ah yes that's great 🔥 
I will do that

File: apps/ledger-live-mobile/src/screens/Portfolio/index.tsx [hunk 5]
```diff
@@ -44,6 +44,8 @@ import CheckTermOfUseUpdate from "../../components/CheckTermOfUseUpdate";
 import TabBarSafeAreaView, {
   TAB_BAR_SAFE_HEIGHT,
 } from "../../components/TabBar/TabBarSafeAreaView";
+import CustomImageBottomModal from "../../components/CustomImage/CustomImageBottomModal";
+import Button from "../../components/Button";
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/screens/Portfolio/index.tsx:48]: Both import are unused ? Is it for future usage ?

Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/screens/Portfolio/index.tsx:48]: ah nope I forgot to remove that

File: apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx (hunk 4 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/src/components/RootNavigator/CustomImageNavigator.tsx:47]: Note: the `as` (ex: `name={ScreenName.DeviceModelSelection as "DeviceModelSelection"}`) is needed as long as `NavigatorName` is an Object and not an enum ... It has been changed by Thomas 1 month ago, but then reverted because one screen was not working (or something, don't really know why it was reverted)

File: apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  const cropperRef = useRef<CropView>(null);
+  const [imageToCrop, setImageToCrop] = useState<
+    (ImageFileUri & Partial<ImageDimensions>) | null
+  >(null);
+  const [rotated, setRotated] = useState(false);
+
+  const { t } = useTranslation();
+
+  const navigation = useNavigation();
+  const { params }: { params: RouteParams } = useRoute();
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx:57]: TS is complaining here - but i guess if you add types to the navigator it will stop complaining ? And i don't know what's best: `useNavigation` and `useRoute`, or pass `navigation` and `route` as props to your component (which is what i did [here](https://github.com/LedgerHQ/ledger-live/blob/aa76c034a6400add49b20d79aba9089681c235f7/apps/ledger-live-mobile/src/screens/SyncOnboarding/BleDevicePairing.tsx#L22) and the props type is defined using `StackScreenProps` [here](https://github.com/LedgerHQ/ledger-live/blob/aa76c034a6400add49b20d79aba9089681c235f7/apps/ledger-live-mobile/src/screens/SyncOnboarding/BleDevicePairing.tsx#L17)) ? With this  props definition, i don't get TS complaining

File: apps/ledger-live-mobile/src/screens/CustomImage/ErrorScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+const Container = styled(SafeAreaView).attrs({
+  edges: ["left", "right", "bottom"],
+})`
+  flex: 1;
+`;
+
+export default () => {
+  const [isModalOpened, setIsModalOpened] = useState(false);
+  const { params }: { params: RouteParams } = useRoute();
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/screens/CustomImage/ErrorScreen.tsx:21]: Same, TS complaining on `params` - i guess it will complain for each screen

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+export async function downloadImageToFile({
+  imageUrl,
+}: ImageUrl): Promise<ImageFileUri> {
+  return downloadImageToFileWithCancellable({ imageUrl })[0];
+}
+
+export async function loadImageToFileWithDimensions(
+  source: Partial<ImageUrl> & Partial<ImageFileUri> & Partial<ImageDimensions>,
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:64]: Is `source: ImageUrl | (ImageFileUri & Partial<ImageDimensions>)` more appropriate ?

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  boxDimensions: ImageDimensions,
+): { height: number; width: number } {
+  const { height: imageHeight, width: imageWidth } = imageDimensions;
+  const { height: boxHeight, width: boxWidth } = boxDimensions;
+  if (imageHeight < boxHeight && imageWidth < boxWidth)
+    return {
+      width: imageWidth,
+      height: imageHeight,
+    };
+  if (imageWidth / imageHeight >= boxWidth / boxHeight) {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:132]: i understand the idea of the ratio, but to guess if the width or the height is the constraint, why not just check `imageWidth > boxWidth` for the width is the constraint (i might miss something ?)

And if `imageHeight === 0` or `boxHeight === 0` the division will fail 🙈

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+  });
+}
+
+export function fitImageContain(
+  imageDimensions: ImageDimensions,
+  boxDimensions: ImageDimensions,
+): { height: number; width: number } {
+  const { height: imageHeight, width: imageWidth } = imageDimensions;
+  const { height: boxHeight, width: boxWidth } = boxDimensions;
+  if (imageHeight < boxHeight && imageWidth < boxWidth)
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:127]: Is it ok if `imageHeight === boxHeight` and `imageWidth === boxWidth` ?

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+          reject(new ImageTooLargeError());
+        } else {
+          reject(new ImageMetadataLoadingError());
+        }
+      },
+    );
+  });
+}
+
+export function fitImageContain(
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:121]: Maybe unit test is overkill, but it could help when dev this kind of util function, because we could test edge cases and check that the result is correct ;) - and maybe we can move it to live-common if one day we do the same for LLD (don't know if it's the plan ?)

File: apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+  );
+
+  /** LOAD SOURCE IMAGE FROM PARAMS */
+  useEffect(() => {
+    loadImageToFileWithDimensions(params)
+      .then(res => {
+        setImageToCrop(res);
+      })
+      .catch(handleError);
+  }, [params, setImageToCrop, handleError]);
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx:74]: From Juan: maybe follow this pattern:
```JS
useEffect(()=>{
  let dead = false;
  someAsyncStuff().then((result)=>{
     if (dead) return;
     // not dead, so do whatever state change
     setStuff(result);
  })
  return ()=>{
    dead = true;
  }
}, [deps]);
```

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts (hunk 9 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:132]: :beaten_trollface:

File: apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        justifyContent="center"
+      >
+        <Flex
+          flex={1}
+          onLayout={onContainerLayout}
+          width="100%"
+          justifyContent="center"
+          alignItems="center"
+        >
+          {(true || false) && imageToCrop ? (
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx:143]: what is this black magic ? 🧙‍♂️

Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx:143]: LOL wtf I do that sometimes when debugging UI 😆 this one is really weird though

Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/screens/CustomImage/Step1Crop.tsx:143]: i really hope you don't find more ridiculous stuff like this

File: apps/ledger-live-mobile/src/components/CustomImage/ImageResizer.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Component to resize an image.
+ *
+ * It:
+ *  - takes as an input an image file URI & target dimensions as an input
+ *  - renders nothing
+ *  - outputs the resulting image as a base64 data URI.
+ *
+ * */
+const ImageResizer: React.FC<Props> = props => {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/ImageResizer.tsx:23]: As this component is not rendering anything, and it has a function (resizing the image), we should transform it in a custom hook ? And call this hook directly in `Step2Preview` screen ? WDYT ? 
And maybe in the future it could be in live-common if we do the same thing in LLD (well it will depend if [ImageManipulator](https://docs.expo.dev/versions/latest/sdk/imagemanipulator/) works on Electron - but that's not for now)

File: apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+  webViewRef: WebView<{}> | null = null;
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.contrast !== this.props.contrast) this.setAndApplyContrast();
+    if (prevProps.imageBase64DataUri !== this.props.imageBase64DataUri)
+      this.computeResult();
+  }
+
+  handleMessage = ({ nativeEvent: { data } }: any) => {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx:60]: i think the type is `WebViewMessageEvent` from [here](https://github.com/react-native-webview/react-native-webview/blob/37c8d34df602d8c693ad261ac99b87b5cd2d4b57/src/WebViewTypes.ts#L1203)

File: apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+
+  function contrastRGB(rgbVal: number, contrastVal: number) {
+    return (rgbVal - 128) * contrastVal + 128;
+  }
+
+  // simutaneously apply grayscale and contrast to the image
+  function applyFilter(
+    imageData: Uint8ClampedArray,
+    contrastAmount: number,
+  ): { imageDataResult: Uint8ClampedArray; hexRawResult: string } {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts:42]: Could be interesting to have some unit test on `applyFilter`: to have examples of what to expect for some random `imageData` array with a contrast. If there is any edge cases ? (imageData all 0 or 1 ?).

File: apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx (hunk 15 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx:60]: And to follow the other "handle" functions, maybe rename it to `handleWebViewMessage` ? (i'm picky 🐩)

File: apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+  requestRawResult = () => {
+    this.injectJavaScript("window.requestRawResult();");
+  };
+
+  computeResult = () => {
+    this.setContrast();
+    this.processImage();
+  };
+
+  handleWebviewLoaded = () => {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx:131]: typo (i guess, compared to `handleWebViewError`): `handleWebViewLoaded`

File: apps/ledger-live-mobile/src/components/CustomImage/ResultDataTester.tsx [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    this.injectJavaScript(
+      `window.reconstructImage(${width}, ${height}, "${hexData}");`,
+    );
+  };
+
+  computeResult = () => {
+    this.processImage();
+  };
+
+  handleWebviewLoaded = () => {
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/ResultDataTester.tsx:81]: small typo, `handleWebViewLoaded`

File: apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx (hunk 15 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/ImageProcessor.tsx:60]: ah yep thanks 👍 the "any" typing is probably a remain of the POC days haha

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts (hunk 9 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:117]: So indeed `0` values with fail haha I will fix that.

Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:117]: Regarding the logic itself, just checking `imageWidth > boxWidth` is not enough, both the width and the height can be the constraint.

The question is which dimension (width or height) is the most constraining.

If it's the width (this case), then we use the width of the box as a basis for resizing (while keeping the original aspect ratio of the image) and then we are sure that the resulting height will fit in the box.

Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:124]: @alexandremgo I added some comments in the code to explicit what is happening.

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts (hunk 11 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:111]: Indeed, I could have unit tested this 😅 
I'll do it if we reuse that for LLD & move it to LLC.

File: apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts (hunk 16 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts:47]: I was planning on rather implementing e2e tests (when they're available on LLM) to directly validate the result on given images.

File: apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts (hunk 9 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/src/components/CustomImage/imageUtils.ts:124]: Love the explanation, thanks :)

File: apps/ledger-live-mobile/src/components/CustomImage/useResizedImage.tsx [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  onError: (error: Error) => void;
+  onResult: (res: ResizeResult) => void;
+};
+
+/**
+ * Component to resize an image.
+ *
+ * It:
+ *  - takes as an input an image file URI & target dimensions as an input
+ *  - renders nothing
```
Comment (alexandremgo) [inline apps/ledger-live-mobile/src/components/CustomImage/useResizedImage.tsx:19]: Nice hook :) just maybe update the comment to say that it's a hook now, so people don't get confused ;)

File: apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts (hunk 16 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/src/components/CustomImage/injectedCode/imageProcessing.ts:47]: yeah good idea :)

**Review by alexandremgo — APPROVED** [review]

File: apps/ledger-live-mobile/src/components/CustomImage/useResizedImage.tsx (hunk 19 above)
Comment (ofreyssinet-ledger) [reply apps/ledger-live-mobile/src/components/CustomImage/useResizedImage.tsx]: thanks! done 👍

**Comment by ofreyssinet-ledger** [comment]
QA OK as per
https://ledgerhq.atlassian.net/browse/FAT-135?focusedCommentId=220655
and the bug mentioned by @alalmi-ledger on this comment is also fixed & mentioned as QA ok here https://ledgerhq.atlassian.net/browse/FAT-366
